clientconn.go

```diff
@@ -840,6 +840,73 @@ func (ac *addrConn) connect() error {
 	ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 	ac.mu.Unlock()
 
+	common := func() bool {
+		ac.mu.Lock()
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return true
+		}
+
+		ac.state = connectivity.TransientFailure
+		ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
+		ac.cc.resolveNow(resolver.ResolveNowOption{})
+		ac.curAddr = resolver.Address{}
+		ac.mu.Unlock()
+
+		if err := ac.resetTransport(); err != nil {
+			ac.mu.Lock()
+			ac.printf("transport exiting: %v", err)
+			ac.mu.Unlock()
+			grpclog.Warningf("grpc: addrConn.transportMonitor exits due to: %v", err)
+			if err != errConnClosing {
+				// Keep this ac in cc.conns, to get the reason it's torn down.
+				ac.tearDown(err)
+			}
+			return true
+		}
+		return false
+	}
+
+	var timer *time.Timer
+	var cdeadline <-chan time.Time
+
+	onDeadline := func() bool {
+		ac.mu.Lock()
+		if ac.backoffDeadline.IsZero() {
+			ac.mu.Unlock()
+			return true
+		}
+		ac.mu.Unlock()
+		err := ac.transport.Close()
+		if err != nil {
+			grpclog.Error(err)
```

Comment:
What's the appropriate thing to do here?

Reply:
Logging seems fine.  There's essentially nothing we can do here to "recover".  The only other option would be `Fatal()`, but `Error()` seems more appropriate since it's non-fatal to fail to close a single Conn, and we want to avoid crashing the program as much as possible.  I believe the only normal way `Conn.Close` would error is if you tried to close the same connection multiple times, anyway, so we should be fine.

Reply:
Right on 👍